### DISCUSSION TITLE
Adding Prisma1 and Prisma2 side-by-side with Breaking Changes Detected

### DISCUSSION THREAD

**Discussion by alan345** [prisma/p1-to-p2-upgrade-path-feedback#9] [Q&A]
## TL;DR 
How can have prisma1 and prisma2 side-by-side if `prisma-upgrade` creates breaking changes?

## Issue
For migrating from prisma1 to prisma2, we decided to keep side by side primsa1 and prisma2.
Running `npx prisma-upgrade` will create 2 breaking changes `Fix one-to-many table relations` and `Fix one-to-one table relations`. In order to make `prisma1` and `prisma2` work together, we will not run the SQL changes.

Now it is time to slowly migrate the codebase to prisma2:

This is the datamodel 1.0 for `User`:

```graphql
type User {
  id: ID! @unique
  lastName: String!
  firstName: String!
  userRoleCompanies: [UserRoleCompanie!]! @relation(name: "UserRoleCompanieRelation", onDelete: CASCADE) 
}

type UserRoleCompanie {
  id: ID! @unique
  isInvitationApproved: Boolean!
  companie: Companie! @relation(name: "UserRoleCompaniesRelation")
}
```

I have to change this query in my codebase: 

```ts
const user = await ctx.db.query.user({ where: { email } },
  `
  {
    id
    firstName
    lastName
    email
    userRoleCompanies(where: { isInvitationApproved: true, companie: { deletedLogically: false } }, first: 1) {
      id
      companieRole
    }
  }
`
)
```

to:

```ts
    const user = await prisma.user
      .findOne({
        where: { email }
      })
      .UserRoleCompanie_UserRoleCompanieRelation({
        where: {
          isInvitationApproved: true,
          Companie: { deletedLogically: false }
        },
        first: 1
      })
```

I got this error: 

[![enter image description here][1]][1]

of course because the new schema (prisma2) is:

```prisma
model UserRoleCompanie {
  id                                       String                         @id
  Companie                                 Companie[]                     @relation("UserRoleCompaniesRelation", references: [id])
}
```

The relationship UserRoleCompanie-Companie is 1-1 in prisma1 and 1-many in prisma2.

If I replace the `schema.prisma`

from

```prisma
Companie                                 Companie[]                     @relation("UserRoleCompaniesRelation", references: [id])
```

to 

```prisma
companie                                 Companie                   @relation(fields: [companiId],  references: [id])
```

I Have the error `Error validating: The argument fields must refer only to existing fields. The following fields do not exist in this model: companiId`. Which make sense because I did not upgrade the database.

How can we move forward without running the SQL to avoid breaking prisma1?

  [1]: https://i.stack.imgur.com/g5lWO.png

**Comment by nikolasburk** [comment]
Hey Alan, thanks a lot for raising this (and sorry for the confusion about the disappearing issue 😅)!

> How can have prisma1 and prisma2 side-by-side if prisma-upgrade creates breaking changes?

Basically you can take the following approach:

- as long as you run side-by-side, you perform all SQL changes that are **note** breaking. that way, you can keep running both versions at the same time while not breaking anything.
- once you want to wipe out the last bits of Prisma 1 from your app, you can run the SQL changes that are breaking since you don't need to worry about Prisma 1 compatibility at that point any more.

Does that make sense? Let me knof if you have any further questions about this :)

**Comment by alan345** [comment]
We have ~600 queries/mutation in our codebase. `await ctx.db.query. ... `

Most of them are nested queries where there is a 1-1 relationship. It means I cannot update these queries/mutation if the database has not been updated with the breaking changes. 

Sorry but I still don't see how we can proceed to a smooth upgrade to prisma2

**Reply by nikolasburk** [reply]
I just wrote you a DM on Slack, let's discuss further details there 🙂
